### DISCUSSION TITLE
A Universal Business Functionality Template Based on NiceGUI？

### DISCUSSION THREAD

**Discussion by v479038280** [zauberzeug/nicegui#3178] [Ideas / Feature Requests]
"Is it appropriate to use NiceGUI for such a [SCADA ](https://github.com/pyscada/PyScada)software? What kind of approach is needed to handle the arrangement of the interface layer, business logic layer, and device communication layer? How to modularize the implementation of alarms and real-time data?  Is there any template for such function?"

**Comment by v479038280** [comment]
maybe something like this 
![微信截图_20240607045239](https://github.com/zauberzeug/nicegui/assets/25096045/44f552af-ab4b-46fa-b3e0-bd4db10072f4)

**Comment by rodja** [comment]
NiceGUI was designed to allow running arbitrary Python in combination with the GUI. In your SCADA software you would simply use NiceGUI for the visualization part and do data acquisition, communication, processing and control with other Python libraries.

**Reply by v479038280** [reply]
What is unclear is how to coordinate the interdependencies between presenting output, retrieving data, and executing computational operations; what guidelines should I follow when designing such workflows?

**Reply by rodja** [reply]
In [RoSys](https://rosys.io/) we control mobile robotic systems and provide the UI with NiceGUI. It's straight forward in the most cases. Simply create your business logic in a `ui.on_startup` callback. Then you can access the created objects in your page-builder functions to retrieve data and save them. See our [Field Friend code](https://github.com/zauberzeug/field_friend/blob/main/main.py) which uses RoSys to build a weeding robot.
